batch/batch/worker/worker.py

```diff
@@ -528,6 +529,81 @@ def __str__(self):
         return f'container {self.job.id}/{self.name}'
 
 
+class JVMProcess:
+    classpath = f'{find_spark_home()}/jars/*:/hail.jar'
+    stack_size = 512 * 1024
+    thread_pool = None
+
+    def __init__(self, job, main_spec):
+        self.job = job
+        self.heap_size = main_spec['memory'] - self.stack_size
+        self.env = {}
+        for var in main_spec.get('env', []):
+            self.env[var['name']] = var['value']
+
+        self.flags = ['-classpath', self.classpath, f'-Xmx{self.heap_size}', f'-Xss{self.stack_size}']
+        self.java_args = main_spec['command']
+
+        self.proc = None
+        self.timing = {}
+        self.state = 'pending'
+        self.log = ''
+
+    async def run(self, worker):
+        log.info(f'running {self}')
+        self.timing['start_time'] = time_msecs()
+        self.proc = await asyncio.create_subprocess_exec(
+            'java',
+            *self.flags,
+            *self.java_args,
+            stdout=asyncio.subprocess.PIPE,
+            stderr=asyncio.subprocess.PIPE,
+            env=self.env)
+        out, err = await self.proc.communicate()
+
+        finish_time = time_msecs()
+        self.timing['finish_time'] = finish_time
+        start_time = self.timing['start_time']
+        self.timing['duration'] = finish_time - start_time
+
+        self.log += 'STDOUT:\n'
+        self.log += out.decode()
+        self.log += '\n\n'
+        self.log += 'STDERR:\n'
+        self.log += err.decode()
+
+        log.info(f'finished {self} with return code {self.proc.returncode}')
+        log.info(f'log {self}: {self.log}')
+
+        await worker.log_store.write_log_file(
+            self.job.format_version, self.job.batch_id,
+            self.job.job_id, self.job.attempt_id, 'main',
+            self.log)
+
+        if self.proc.returncode == 0:
+            self.state = 'succeeded'
+        else:
+            self.state = 'error'
```

Comment:
non-zero exit code is `'failed'` whereas something like python can't open a subprocess is `'error'`.